orientdb/src/main/scala/akka/stream/alpakka/orientdb/OrientDBFlowStage.scala

```diff
@@ -0,0 +1,211 @@
+
+import scala.collection.mutable
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+final case class IncomingMessage[T](oDocument: T)
+
+private[orientdb] class OrientDBFlowStage[T, R](
+    className: String,
+    settings: OrientDBSinkSettings,
+    pusher: Seq[IncomingMessage[T]] => R,
+    clazz: Option[Class[T]]
+) extends GraphStage[FlowShape[IncomingMessage[T], Future[R]]] {
+
+  private val in = Inlet[IncomingMessage[T]]("messages")
+  private val out = Outlet[Future[R]]("failed")
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      private var state: State = Idle
+      private val queue = new mutable.Queue[IncomingMessage[T]]()
+      private val failureHandler = getAsyncCallback[(Seq[IncomingMessage[T]], Throwable)](handleFailure)
+      private val responseHandler = getAsyncCallback[(Seq[IncomingMessage[T]], Option[String])](handleResponse)
+      private var failedMessages: Seq[IncomingMessage[T]] = Nil
+      private var retryCount: Int = 0
+
+      override def preStart(): Unit =
+        pull(in)
+
+      private def tryPull(): Unit =
+        if (queue.size < settings.bufferSize && !isClosed(in) && !hasBeenPulled(in)) {
+          pull(in)
+        }
+
+      override def onTimer(timerKey: Any): Unit = {
+        sendOSQLBulkInsertRequest(failedMessages)
+        failedMessages = Nil
+      }
+
+      private def handleFailure(args: (Seq[IncomingMessage[T]], Throwable)): Unit = {
+        val (messages, exception) = args
+        if (retryCount >= settings.maxRetry) {
+          failStage(exception)
+        } else {
+          retryCount = retryCount + 1
+          failedMessages = messages
+          scheduleOnce(NotUsed, settings.retryInterval.millis)
+        }
+      }
+
+      private def handleSuccess(): Unit =
+        completeStage()
+
+      private def handleResponse(args: (Seq[IncomingMessage[T]], Option[String])): Unit = {
+        retryCount = 0
+        val (messages, error) = args
+
+        val failedMessages = messages.flatMap {
+          case message =>
+            if (error.isEmpty) {
+              None
+            } else {
+              Some(message)
+            }
+        }
+
+        val nextMessages = (1 to settings.bufferSize).flatMap { _ =>
+          queue.dequeueFirst(_ => true)
+        }
+
+        if (nextMessages.isEmpty) {
+          state match {
+            case Finished => handleSuccess()
+            case _ => state = Idle
+          }
+        } else {
+          sendOSQLBulkInsertRequest(nextMessages)
+        }
+
+        push(out, Future.successful(pusher(failedMessages)))
+      }
+
+      private def sendOSQLBulkInsertRequest(messages: Seq[IncomingMessage[T]]): Unit = {
+        val oDatabase = new OPartitionedDatabasePool(settings.dbUrl, settings.username, settings.password)
```

Comment:
Shouldn't this (creating & getting the connection) be done in the `preStart` method?
And similar, having a `postStop` method to clean up stuff?